### PR TITLE
Block library: reusable caption component util

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Prep work for #43204

This is the same logic used by the Image, Gallery, Audio, and Video blocks. Let's make it reusable because it's a pain to change every instance.

Note that I've also optimised these `useEffect` and `useCallback` calls. Previously these callbacks would get called whenever you type in the caption, now they're only called if emptiness changes.

Additionally emptiness checks check the length.

## Why?

See https://github.com/WordPress/gutenberg/pull/43204#discussion_r1406925868

### CODE REVIEW

File: packages/block-library/src/audio/edit.js [hunk 1]
```diff
@@ -73,30 +67,6 @@ function AudioEdit( {
...
-			}
-		},
-		[ caption ]
-	);
-
-	useEffect( () => {
-		if ( ! isSelected && ! caption ) {
-			setShowCaption( false );
-		}
-	}, [ isSelected, caption ] );
```
Comment (ellatrix) [inline packages/block-library/src/audio/edit.js:98]: Note that I've also optimised these `useEffect` and `useCallback` calls. Previously these callbacks would get called whenever you type in the caption, now they're only called if emptiness changes.

**Review by dmsnell — COMMENTED** [review]
I definitely don't feel qualified to properly review this, but it looks really nice, and a good way to reduce our overall code overhead.

**Review by dmsnell — APPROVED** [review]
Let's get it in use/testing!

File: packages/block-library/src/utils/caption.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	RichText,
+	BlockControls,
+	__experimentalGetElementClassName,
+} from '@wordpress/block-editor';
+import { ToolbarButton } from '@wordpress/components';
+import { caption as captionIcon } from '@wordpress/icons';
+import { createBlock, getDefaultBlockName } from '@wordpress/blocks';
+
+export function Caption( {
+	key = 'caption',
```
Comment (jsnajdr) [inline packages/block-library/src/utils/caption.js:22]: Can this really be used to make `<Caption>` edit an attribute other than `caption`? `key` or `ref` is a reserved prop, React won't really pass it down to the component. Let's find a better name for the prop.

File: packages/block-library/src/utils/caption.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+						className={ classnames(
+							className,
+							__experimentalGetElementClassName( 'caption' )
+						) }
+						ref={ ref }
+						aria-label={ label }
+						placeholder={ placeholder }
+						value={ caption }
+						onChange={ ( value ) =>
+							setAttributes( { caption: value } )
```
Comment (jsnajdr) [inline packages/block-library/src/utils/caption.js:96]: Should be `setAttributes( { [ key ]: value } )`.

File: packages/block-library/src/utils/caption.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-library/src/utils/caption.js:22]: It was just an unfinished idea. See #59073

Comment (jsnajdr) [reply packages/block-library/src/utils/caption.js:22]: Nice, I see it's all fixed there :+1:

**Pull request merged**